optimistix/_solver/quasi_newton.py

```diff
@@ -115,12 +78,615 @@ def __call__(
+                lambda y: _lbfgs_hessian_operator_fn(y, state),
+                jax.eval_shape(lambda: y),
+                tags=lx.positive_semidefinite_tag,
+            )
+            f_info = FunctionInfo.EvalGradHessian(f, grad, operator)  # pyright: ignore
+            return f_info, state  # pyright: ignore  # TODO
+
+    def _no_update(self, inner, grad_diff, y_diff, f_info, hessian_update_state):
+        if isinstance(f_info, FunctionInfo.EvalGradHessianInv):
+            operator = f_info.hessian_inv
+        else:
+            operator = f_info.hessian
+        return eqx.filter(operator, eqx.is_array), hessian_update_state
+
+    def _update(self, inner, grad_diff, y_diff, f_info, state):
+        updated_y_diff_history = jtu.tree_map(
+            lambda x, z: x.at[state.index_start % self.history_length].set(z),
+            state.y_diff_history,
+            y_diff,
+        )
+        updated_grad_diff_history = jtu.tree_map(
+            lambda x, z: x.at[state.index_start % self.history_length].set(z),
+            state.grad_diff_history,
+            grad_diff,
+        )
+
+        # Explicitly hit different code paths for the updates to the approximate inverse
+        # or regular Hessian. This is to make pyright happy, as it can then verify that
+        # the type of the state matches the expected type of the returned FunctionInfo
+        # in __call__ below.
+        # update conditionally
+        if self.use_inverse:
+            updated_inner_history = state.inner_history.at[
+                state.index_start % self.history_length
+            ].set(1.0 / inner)
+            updated_state = _LBFGSInverseHessianUpdateState(
+                index_start=state.index_start + 1,
+                y_diff_history=updated_y_diff_history,
+                grad_diff_history=updated_grad_diff_history,
+                inner_history=updated_inner_history,
+            )
+            operator = lx.FunctionLinearOperator(
+                lambda y: _lbfgs_inverse_hessian_operator_fn(y, updated_state),
+                jax.eval_shape(lambda: y_diff),
+                tags=lx.positive_semidefinite_tag,
+            )
+
+            # Only return the dynamic part of the operator, keep jaxpr across iterations
+            return eqx.filter(operator, eqx.is_array), updated_state
+
+        else:
+            # if not self.use_inverse:
+            # roll the lower-triangular matrix, sort the inner product
+            # according to history, set the last row as the new inner product.
+            y_diff_grad_diff_cross_inner = roll_low_triangular_matrix(
+                state.y_diff_grad_diff_cross_inner,
+                v_tree_dot(state.grad_diff_history, y_diff),
+                state.index_start,
+            )
+
+            y_diff_grad_diff_inner = jnp.roll(
+                state.y_diff_grad_diff_inner,
+                -1 * (state.index_start > self.history_length - 1).astype(int),
```

Comment:
Avoid `.astype(int)` as this produces platform-dependent results. Probably prefer an explicit `np.int32`.